ark/utils/spatial_analysis_utils.py

```diff
@@ -266,6 +251,150 @@ def compute_close_cell_num_random(marker_nums, dist_mat, dist_lim, bootstrap_num
+            the name of the column in current_fov_data which contains the FlowSOM ID
+
+    Returns:
+        numpy.ndarray:
+            Large matrix of random positive marker counts for every permutation in the bootstrap
+    """
+
+    # TODO: basic checking to see if all specified cell_type_facets
+    # exist in the FlowSOM ID col of current_fov_data
+
+    # TODO: basic checking to see if cell_type_rand percents sum up to less than 1
+
+    # Create close_num_rand
+    close_num_rand = np.zeros((
+        len(marker_nums), len(marker_nums), bootstrap_num), dtype='int')
+
+    dist_mat_bin = (dist_mat < dist_lim).astype(np.int8)
+
+    # create a dictionary to store information about each cell type
+    # copy over the percent value, will be filled with a lot more goodies
+    # and get the indices that corrrespond to each cell type in current_fov_data
+    cell_type_data = dict(zip(cell_type_rand.keys(), {'percent': cell_type_rand.values()}))
+    cell_type_data = {str(m): {'percent': cell_type_rand[m],
+                               'indices': current_fov_data[current_fov_data[cell_type_col] == m]
+                               .index.values}
+                      for m in cell_type_rand}
+
+    # TODO: trying to use Adam's optimization, not currently working
+    # cell_type_data = dict(zip(cell_type_rand.keys(),
+    #                           {'percent': cell_type_rand.values(),
+    #                            'indices': current_fov_data[cell_type_col] ==
+    #                                       np.expand_dims(np.array(list(cell_type_rand.keys())),
+    #                                                      axis=1)}))
+
+    # the else column will basically be the inverse of everything else
+    # percentage is the 1 - sum(cell_type_rand.values())
+    # indices are whatever rows do not correspond to a FlowSOM ID specified in cell_type_rand
+
+    cell_type_data['else'] = {
+        'percent': 1 - sum(cell_type_rand.values()),
+        'indices': current_fov_data[~current_fov_data[cell_type_col].isin(
+            list(cell_type_rand.keys()))].index.values
+    }
+
+    for j, m1n in enumerate(marker_nums):
+        # recalculate markerposinds, we need this to get the actual marker_counts per cell type
+        marker_col = current_fov_channel_data.columns[j]
+        marker_pos_select = current_fov_channel_data[marker_col] > thresh_vec[j]
+        marker_pos_inds = current_fov_channel_data[marker_pos_select].index.values
+
+        # now take the intersection between marker_pos_inds and the indices associated
+        # with each cell type, we'll need this for both the total marker_inds per cell_type
+        # and the actual indices needed to subset the distance matrix
+        for cell_type in cell_type_data:
+            cell_type_data[str(cell_type)]['marker_inds'] = \
+                list(set(marker_pos_inds).intersection(set(
+                    cell_type_data[str(cell_type)]['indices'])))
```

Comment:
OK, so indices is not actually a bool array, but rather the list of indices in current_fov_data that are associated with the cell_type in question. marker_pos_inds, on the other hand, are the indices associated with the marker num in question. marker_pos_select is just a helper indexing variable into current_fov_channel_data (it would be too long if I used a one-liner for that).

Perhaps I should add a comment clarifying this in my next commit.